keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,374 @@
+
+## Related issues
+
+- [Hardware topology awareness at node level (including NUMA)][numa-issue]
+- [Support Container Isolation of Hugepages][hugepage-issue]
+- [Support isolating memory to single NUMA node][memory-issue]
+
+## Goals
+
+- To guarantee isolation of memory and hugepages to single NUMA node for containers in the Guaranteed Pod.
+
+- To provide memory and hugepages topology hint to the Topology Manager. It will allow Topology Manager to allocate memory, hugepages, CPUs, and PCI devices on the same NUMA node.
+
+## Non-Goals
+
+- Supporting cross NUMA affinity is out of scope.
+
+- Isolating memory & hugepage to a NUMA node for multiple containers is out of scope.
+
+- Updating scheduler, pod spec is out of scope at this point.
+
+- This proposal only focus on Linux based systems.
+
+## User Stories
+
+### Story 1 : Networking Acceleration using DPDK
+
+- The system such as real-time trading system and 5G system, which requires networking acceleration, uses DPDK to guarantee low latency of packet processing. DPDK(Data Plane Development Kit) is set of libraries to accelerate packet processing on userspace. DPDK requires dedicated resources(such as exclusive cpu usage, hugepages and DPDK compatile Network Interface Card) and alignment of resources to sigle NUMA node to accelerate its performance. For this reason, there should be a way to guarantee a resource reservation of memory and hugepage as well as other computing resources (i.e. cpu core , NIC) from a single NUMA node for DPDK-based containers.
+
+### Story 2 : Database
+
+- Databases(Oracle, PostgreSQL, MySQL, MongoDB, SAP) consumes massive Memory and Hugepages, in order to reduce memory access latency and improve performance dramatically resources(CPU, Memory, Hugepages, I/O Devices) should be aligned.
+
+# Proposal
+
+## Proposed Changes
+
+### New Component: Memory Manager
+
+Memory Manager is a new component of Kubelet and a new hint provider of Topology Manager for memory and hugepage. It provides NUMA node affinity to the Topology Manager to guarantee NUMA aware memory consumption for a container in the Guaranteed QoS Pod.
+
+When Guaranteed QoS Pod admission is requested to kubelet, Topology Manager asks NUMA node affinity of memory and hugepage to Memory Manager. Then Memory Manager calculates NUMA node affinity based on internal database(`Memory Map`) and sends the affinity back to the Topology Manager to help the Topology Manager to figure out appropriate NUMA node where adequate resources are available for a given container.
+
+Note:
+- Memory Manager only support single NUMA node affinity, because Linux kernel doesn`t have such mechanism to reserve memory or set memory limit per NUMA node.
```

Comment:
Can you elaborate on this Linux kernel restriction a little bit more?

Reply:
I'm thinking about it to add kernel restriction as a part of the appendix.
Just like `How kernel allocates memory and bind physical pages`.